File: app/src/main/java/ru/mrlargha/stemobile/data/Result.java
package ru.mrlargha.stemobile.data;

import org.jetbrains.annotations.NotNull;

/**
 * Результат выполнения запроса
 *
 * @param <T> тип ответа сервера
 */
public class Result<T> {
    @NotNull
    @Override
    public String toString() {
        if (this instanceof Result.Success) {
            Result.Success success = (Result.Success) this;
            return "Success[data=" + success.getData().toString() + "]";
        } else if (this instanceof Result.Error) {
            Result.Error error = (Result.Error) this;
            return "Error[exception=" + error.getErrorString() + "]";
        }
        return "";
    }

    // Success sub-class
    public final static class Success<T> extends Result {
        private final T data;

        Success(T data) {
            this.data = data;
        }

        public T getData() {
            return this.data;
        }
    }

    // Error sub-class
    public final static class Error extends Result {
        private final String errorString;

        public Error(String errorString) {
            this.errorString = errorString;
        }

        public String getErrorString() {
            return this.errorString;
        }
    }
}

File: app/src/main/java/ru/mrlargha/stemobile/data/Result.kt
package ru.mrlargha.stemobile.data

open class Result<T> {
    override fun toString(): String {
        return when (this) {
            is Success<*> -> "Success[data=" + this.data.toString() + "]"
            is Error -> "Error[exception=" + this.errorString + "]"
            else -> "Undefined type!"
        }
    }

    // Success sub-class
    class Success<T>(val data: T) : Result<Any?>()

    // Error sub-class
    class Error(val errorString: String) : Result<Any?>()
}